### PR TITLE
Drools 2909 - Dynamically populate Right panel

### PR DESCRIPTION
**Pull request by gitgabrio**
Scope of this PR is to dynamically populate the right panel with properties (FactModels) retrieved from DMO.
Properties will be lazy-loaded, i.e. nested property (if present) will be retrieved and shown on property' row  expand.
No other behaviour will be implemented.
@kkufova @Rikkola

### CODE REVIEW

**Comment by kiereleaseuser3** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

**Comment by Rikkola** [comment]
ok to test

**Comment by kiereleaseuser2** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

**Comment by kiereleaseuser2** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        listGroupItemHeader.removeClassName("list-view-pf-expand-active");
+        listGroupItemContainer.addClassName("hidden");
+        faAngleRight.removeClassName("fa-angle-down");
+    }
+
+    @Override
+    public void expandRow() {
+        listGroupItemHeader.addClassName("list-view-pf-expand-active");
+        listGroupItemContainer.removeClassName("hidden");
+        faAngleRight.addClassName("fa-angle-down");
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java]: I think we could introduce constants for: "list-view-pf-expand-active", "hidden" and "fa-angle-down".

Comment (gitgabrio) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java]: Fixed with
[DROOLS-2909] Moved repeated String to constants

**Comment by Rikkola** [comment]
@kkufova Did you reopen the Scenario Simulation editor after creating or deleting data objects?

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/FieldItemPresenter.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public LIElement getLIElement(String factName, String fieldName, String className) {
+        FieldItemView fieldItemView = getFieldItemView();
+        fieldItemView.setFieldData(factName, fieldName, className);
+        fieldItemView.setPresenter(this);
+        LIElement toReturn = fieldItemView.getLIElement();
+        return toReturn;
+    }
+
+    protected FieldItemView getFieldItemView() {  // This is needed for test because Mockito can not mock Instance
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/FieldItemPresenter.java]: Why is there a need for the instance. Isn't it just possible to inject the view?

Comment (Rikkola) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/FieldItemPresenter.java]: Ok so this is a provider for list items. Now I understand it. Maybe we should call it FieldItemPresenterProvider or something similar.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemPresenter.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                    populateListGroupItemView(listGroupItemView, listGroupItemView.getFactName(), factModelTree);
+                    listGroupItemView.setToExpand(false);
+                }
+                listGroupItemView.expandRow();
+            }
+        }
+    }
+
+    protected ListGroupItemView getListGroupItemView() {  // This is needed for test because Mockito can not mock Instance
+        return instance.get();
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemPresenter.java]: Again if this just provides the Presenters rename to PresenterProvider

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    DivElement fullClassName = Document.get().createDivElement();
+
+    @DataField("factProperties")
+    UListElement factProperties = Document.get().createULElement();
+
+    Presenter presenter;
+
+    String factName;
+
+    String factType;
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java]: It is a bit dangerous to have any other fields than UI components and the presenter. This leaks the model into the view and will likely have bugs due to out of sync issues with Model, view and view fields. The data is already set into the widgets, if you need to store the fields keep them in the Presenter or in the Model

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelView.java [hunk 5]
```diff
@@ -16,26 +16,35 @@
...
-    void showCheatSheetTab();
+    void showClearButton();
 
-    void hideEditorTab();
+    DivElement getListContainer();
 
     interface Presenter {
 
-        void onEditorTabActivated();
+        void clearSearch();
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelView.java]: To clearly point out these methods are used by the view please use the format onWhatActionIsHapening-format.

The View should not have actions on the Presenter, the Presenter should react on changes in the View.

I took a look at the methods, they are good, just the phrasing could be different.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/FieldItemPresenterTest.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(GwtMockitoTestRunner.class)
+public class FieldItemPresenterTest extends AbstractRightPanelTest {
+
+    @Mock
+        protected FieldItemView mockFieldItemView;
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/FieldItemPresenterTest.java]: Minor formatting issue.

**Comment by gitgabrio** [comment]
@kkufova About the "status" (searchfield/name field) this has been fixed with
[DROOLS-2909] Clearing state on RightPanel hide/close

For the overall data objects, I think we'll keep it as it is, since it is consistent with the other editors, as 
per @Rikkola suggestion.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemPresenter.java (hunk 3 above)
Comment (gitgabrio) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemPresenter.java]: I have implemented a ViewsProvider to separate "instantiation" logic from presenting one

[DROOLS-2909] Implemented ViewsProvider to clearly separate "provider" logic from presenter one

**Comment by kkufova** [comment]
@gitgabrio, I checked the UI again and the following still does not work:

1. Create a data object.
2. Create a scenario. Close the scenario.
3. Remove the data object.
4. Open the scenario. -> The removed data object is still there. Even if you remove the scenario completely and create a new one, the data object will be there (when there is no data object at all in the Library).

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelView.java (hunk 5 above)
Comment (gitgabrio) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelView.java]: Fixed with
[DROOLS-2909] Renaming presenters' methods to follow common usage

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java (hunk 4 above)
Comment (gitgabrio) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/ListGroupItemViewImpl.java]: @Rikkola As we talked about, for that specific case probably it makes sense to keep those tow properties. In general, I agree to have a separated model

**Comment by gitgabrio** [comment]
@kkufova @Rikkola 
I've implemented the "simple" search with
[DROOLS-2909] Implemented basic "filter" search.
From my point of you it is "done" and could be tested again.

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelPresenter.java [hunk 7]
```diff
@@ -66,22 +79,71 @@ public String getTitle() {
...
+    }
+
+    @Override
+    public void onShowClearButton() {
+        view.showClearButton();
+    }
+
+    @Override
+    public void onSearchedEvent(String search) {
+        DivElement listContainer = view.getListContainer();
```
Comment (Rikkola) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelPresenter.java]: You might not want to leak the UI elements into the presenter here. If there are 500+ classes the UI gets so slow that this code will not run properly. I would do the search on the DMO and then fill the UI based on that.

And check with UX what to do if there is thousands of classes in the list. Should we just show the first 20 or what.

Comment (Rikkola) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelPresenter.java]: For now I would just make this work against DMO and make a ticket for the bloated list. Or we can leave it as it is for now and make a ticket for both.

**Comment by gitgabrio** [comment]
@kkufova 
Fixed the "last item" bug with
[DROOLS-2909] Cleaning fact items on RightPanel lost focus

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelPresenter.java (hunk 7 above)
Comment (gitgabrio) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/RightPanelPresenter.java]: @Rikkola good shot, thx!
Fixed with
[DROOLS-2909] Changed search logic. Filtering original fact map and set view's items with filtered entries

**Comment by gitgabrio** [comment]
@kkufova @Rikkola 
I've fixed with
[DROOLS-2909] Fixing hide/open callbacks. Implement search on search button. Fixed RightPanel status

**Comment by kkufova** [comment]
jenkins please retest this

**Comment by kkufova** [comment]
@gitgabrio, the "last item" bug is here again. :-(

**Comment by kiereleaseuser3** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

**Comment by kiereleaseuser3** [comment]
Can one of the admins verify this PR? Comment with 'ok to test' to start the build.

**Comment by Rikkola** [comment]
ok to test

**Comment by gitgabrio** [comment]
@kkufova @kkufova 
Refactored navigation and fixing "last item" issue with:

[DROOLS-2909] Fixing PlaceManager management with PathPlaceRequest. Fixing "last item" issue

@kkufova if still bugged pls setup a BJ for screen share

**Comment by gitgabrio** [comment]
@manstis 
Could you merge this, please? Thx

**Pull request merged**